### ISSUE TITLE
parseLogoutResponse destructs a possibly undefined value while extracting LogoutResponse

### ISSUE THREAD

**Issue by keptt** [boxyhq/saml20#744]
### Problem description

It seems that [the following line](https://github.com/boxyhq/saml20/blob/fdda4acbb5aafdfb3f3c49bbc22e2114b59975f4/lib/logout.ts#L18C27-L18C45) in the `parseLogoutResponse` function:
```
(err: Error | null, { LogoutResponse }) => {
```
leads to the following error:
```
TypeError: Cannot destructure property 'LogoutResponse' of 'undefined' as it is undefined.
```
In case `err` argument is not null.
This overshadows the actual error leading to confusion.

In my specific case, I get the following first argument:
```
err.toString()
'Error: Non-whitespace before first tag.
Line: 0
Column: 1
Char: }'
```
and second argument which is being destructed is undefined.

For me this happens while parsing this response from [saml-mock](https://samlmock.dev/) SLO endpoint:
```
fZFLa8MwEIT/itHd0cPxoyI2lOZiSC9NyKGXsJHXjaksGa0M/fklDqUplByXnZ2Pmd0QjHbSO//h5/iGNHlHmHyN1pFeVjWbg9MeaCDtYETS0ej98+tOq5XQU/DRG29Z0m5rdsqhl12VGWWgX2MG57woDOQl5pWUT2XJktb9QA6+ZieRPXWI0uR9hQZVJs9lxpIjBhq8q5laCZa0RDO2jiK4WDMl1DqVKpXFQVZaSK3KVVYU7yzZIsXBQVwuLzFOmnPrDdiLp6gzIQSHaeAe5njh0EcMqV1ipwG7IaCJrNlcM+uFGO5aeFwCEGG4YllzFaejN58bfud0s530PkKc6e/04jtMjmBnfAyhRa33szFIxHhzI/ya8v8+2XwD
```

### Proposed solution

Destruct only after checking the error.
Instead of:
```
(err: Error | null, { LogoutResponse }) => {
        if (err) {
          reject(err);
          return;
        }
        ...
```
do:
```
(err: Error | null, parsed) => {
        if (err) {
          reject(err);
          return;
        }
        const { LogoutResponse } = parsed;
        ...
```

If this is acknowledged as an issue, I can open a corresponding PR with a fix.

**Comment by niwsa** [comment]
@keptt Thanks for reporting this issue. We will have a look and let you know soon.

**Comment by keptt** [comment]
Thanks @niwsa 

I also have a follow-up, which probably deserves its own issue though.
I assumed that the SLO response I provided above was malformed because when decoded from base64, it was giving me some gibberish instead of a valid XML which was causing sax parser to fail.
However, I went to [samltool.io](samltool.io) and it decoded this response just fine. The reason is that the response is not only base64 encoded but also compressed. So, in order to decompress it I used the following code snippet:

```javascript
const { inflateRawSync } = require('zlib');

samlResponse = 'fZFLa8MwEIT/itHd0cPxoyI2lOZiSC9NyKGXsJHXjaksGa0M/fklDqUplByXnZ2Pmd0QjHbSO//h5/iGNHlHmHyN1pFeVjWbg9MeaCDtYETS0ej98+tOq5XQU/DRG29Z0m5rdsqhl12VGWWgX2MG57woDOQl5pWUT2XJktb9QA6+ZieRPXWI0uR9hQZVJs9lxpIjBhq8q5laCZa0RDO2jiK4WDMl1DqVKpXFQVZaSK3KVVYU7yzZIsXBQVwuLzFOmnPrDdiLp6gzIQSHaeAe5njh0EcMqV1ipwG7IaCJrNlcM+uFGO5aeFwCEGG4YllzFaejN58bfud0s530PkKc6e/04jtMjmBnfAyhRa33szFIxHhzI/ya8v8+2XwD'
const compressed = Buffer.from(samlResponse, 'base64'); // looks like gibberish here
const decompressed = inflateRawSync(compressed).toString(); // and here it finally produces a valid xml
```

I am not sure if your parser is supposed to handle such cases, but it would be very nice, in my opinion.
However, the problem with decompression is that it throws an error if the string is not compressed to begin with. So we would first have to decode and trim a string, and then check if it looks like a valid XML before decompressing. Something like this:

```javascript
...
let result = Buffer.from(samlResponse, 'base64').toString('utf8');
if (!result.trim().startsWith('<')) {
    result =  inflateRawSync(Buffer.from(samlResponse, 'base64')).toString();
}
...
```
This also leads me to believe that there might be other places where such checks and decompression could be appropriate but I haven't looked into those yet.

But as mentioned, this is a different thing from the reported problem, so I can open a separate issue for this if you say that this is something in the scope of the library.

**Comment by keptt** [comment]
also, my previous comment is more relevant for jackson repo, as SAML response is [decoded there ](https://github.com/boxyhq/jackson/blob/feaa341c2075346984d1bd93fbadbec8d1501354/npm/src/controller/logout.ts#L107C5-L107C72) before being passed to this library

**Comment by niwsa** [comment]
@keptt Feel free to open a PR for the `parseLogoutResponse` destructuring bug. 
For the SLO response handling you are right, the issue doesn't belong here. We will need to see what the RFC says with respect to inflate and deflate (will revert back).

**Comment by deepakprabhakara** [comment]
@keptt We seem to be handling it there already but can check this case as well in case we missed it.
